File: greedy/island_connect.py
def union(a, b, node):
    parents_a = find(a, node)
    parents_b = find(b, node)

    if parents_a == parents_b:
        return False

    node[parents_b] = parents_a

    return True


def find(a, node):
    if a != node[a]:
        node[a] = find(node[a], node)
    return node[a]


def solution(n, costs):
    answer = 0
    costs = sorted(costs, key=lambda x: x[2])
    node = [i for i in range(n)]

    for x, y, cost in costs:

        if union(x, y, node):
            answer += cost
    return answer

File: greedy/island_connect.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;
int node[101];

int findParents(int a) {
    if(a != node[a]){
        node[a] = findParents(node[a]);
    }
    return node[a];
}
bool _union(int a, int b){
    a = findParents(a);
    b = findParents(b);
    
    if (a == b) return false;
    
    node[b] = a;
    
    return true;
}

bool comp(vector<int> a, vector<int> b) {
    return a[2] < b[2];
}

int solution(int n, vector<vector<int>> costs) {
    int answer = 0;
    sort(costs.begin(),costs.end(), comp);
    for(int i=0; i<n; i++){
        node[i] = i;
    }
    for(int i=0; i<costs.size(); i++) {
        if(_union(costs[i][0],costs[i][1])){
            answer += costs[i][2];
        }
    }
    return answer;
}